### PR TITLE
bugfix: reverse merkle path to match other structures

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

The store builds the path from root to leaf, this updates the code to return a path from leaf to root, as it is done by the other structures.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: src/merkle/store.rs [hunk 1]
```diff
@@ -220,8 +220,6 @@ impl MerkleStore {
             self.add_merkle_path(index_value, node, path)?;
         }
 
-        // Returns the parent of the last paths (assumes all paths have the same parent) or empty
-        // The length of unique_roots is checked above, so this wont panic
```
Comment (hackaugusto) [inline src/merkle/store.rs:224]: This was changed on the previous PR, I forgot to remove the comment

File: src/merkle/store.rs [hunk 2]
```diff
@@ -265,7 +263,7 @@ impl MerkleStore {
         let mut hash: RpoDigest = root.into();
         let mut path = Vec::new();
         let node = RpoDigest::default();
-        for bit in index.bit_iterator() {
+        for bit in index.bit_iterator().rev() {
```
Comment (hackaugusto) [inline src/merkle/store.rs:266]: now it goes from the leaft to the root

File: src/merkle/store.rs [hunk 3]
```diff
@@ -280,6 +278,7 @@ impl MerkleStore {
             }
         }
 
+        path.reverse();
```
Comment (hackaugusto) [inline src/merkle/store.rs:281]: and this reverse the path to match leaf to root

File: src/merkle/store.rs [hunk 4]
```diff
@@ -469,27 +620,109 @@ mod test {
...
+        let depth = 3;
+        let set = MerklePathSet::new(depth).with_paths(paths).unwrap();
 
+        // STORE LEAVES ARE CORRECT ==============================================================
+        // checks the leaves in the store corresponds to the expected values
         assert_eq!(
-            set.get_node(NodeIndex::new(3, 0)).unwrap(),
-            store.get_node(set.root(), NodeIndex::new(2, 0b00)).unwrap(),
+            store
+                .get_node(set.root(), NodeIndex::new(set.depth() - 1, 0))
```
Comment (hackaugusto) [inline src/merkle/store.rs:635]: I think we should fix this inconsistency, comments?

File: src/merkle/store.rs [hunk 5]
```diff
@@ -280,6 +278,7 @@ impl MerkleStore {
             }
         }
 
+        path.reverse();
         Ok((node.into(), MerklePath::new(path)))
```
Comment (vlopes11) [inline src/merkle/store.rs:282]: We return `node.into()` here, but it is always set to `RpoDigest::default()`. What is the original intent? To be the leaf or the target root?

**Review by vlopes11 — APPROVED** [review]
Looks good! There is also something with the extra returned word from `get_path`, but maybe it is unrelated to this PR

File: src/merkle/store.rs (hunk 5 above)
Comment (hackaugusto) [reply src/merkle/store.rs:282]: the path doesn't include the root, that should have been it, the shadowing of the `node` variable was a bug.

File: src/merkle/store.rs (hunk 4 above)
Comment (bobbinth) [reply src/merkle/store.rs:635]: Is the issue here with `MerklePathSet`?

File: src/merkle/store.rs [hunk 6]
```diff
@@ -110,20 +110,39 @@ impl MerkleStore {
...
     ///
     /// This will return `NodeNotInStorage` if the element is not present in the store.
-    pub fn get_path(&self, root: Word, mut index: NodeIndex) -> Result<MerklePath, MerkleError> {
-        let mut path = Vec::with_capacity(index.depth().saturating_sub(1) as usize);
-        while index.depth() > 0 {
-            let sibling = index.sibling();
-            index.move_up();
-            let node = self.get_node(root, sibling)?;
-            path.push(node);
+    pub fn get_path(
```
Comment (vlopes11) [inline src/merkle/store.rs:116]: Would be better to add a documentation what is the first word of the tuple

File: src/merkle/store.rs [hunk 7]
```diff
@@ -110,20 +110,39 @@ impl MerkleStore {
...
+        &self,
+        root: Word,
+        index: NodeIndex,
+    ) -> Result<(Word, MerklePath), MerkleError> {
+        let mut hash: RpoDigest = root.into();
+        let mut path = Vec::new();
```
Comment (vlopes11) [inline src/merkle/store.rs:122]: optional: this vec can be created `with_capacity` `index.depth()`

**Review by vlopes11 — APPROVED** [review]
Looks good!

File: src/merkle/store.rs (hunk 4 above)
Comment (hackaugusto) [reply src/merkle/store.rs:713]: It seems that `MerklePathSet` is 1-indexed, wereas the other structures are 0-indexed.

In this test `mtree.depth()` is 2, and `set.depth()` is 3 because of this `+1`:

https://github.com/0xPolygonMiden/crypto/blob/962a07292f468e872f7453f13d9df28f2a78291c/src/merkle/path_set.rs#L136

File: src/merkle/store.rs [hunk 8]
```diff
@@ -284,16 +301,18 @@ impl MerkleStore {
...
         index: NodeIndex,
         value: Word,
     ) -> Result<Word, MerkleError> {
-        let current_node = self.get_node(root, index)?;
-        let path = self.get_path(root, index)?;
-        if current_node != value {
-            self.add_merkle_path(index.value(), value, path)
+        let result = self.get_path(root, index)?;
+        if result.0 != value {
+            self.add_merkle_path(index.value(), result.0, result.1)
```
Comment (bobbinth) [inline src/merkle/store.rs:315]: Let's add doc comments to this function.

Also, shouldn't line 315 be:
```Rust
self.add_merkle_path(index.value(), value, result.1)
```
If so, should we add a test to make sure this is checked?

File: src/merkle/store.rs (hunk 6 above)
Comment (hackaugusto) [reply src/merkle/store.rs:116]: My idea was to add the return struct that we discussed on the original PR. But I can add docs too.

File: src/merkle/store.rs [hunk 9]
```diff
@@ -346,49 +365,145 @@ mod test {
...
+        // STORE LEAVES ARE CORRECT ==============================================================
+        // checks the leaves in the store corresponds to the expected values
+        assert_eq!(
             store
                 .get_node(mtree.root(), NodeIndex::new(mtree.depth(), 0))
-                .is_ok(),
+                .unwrap(),
+            LEAVES4[0],
             "node 0 must be in the tree"
         );
```
Comment (bobbinth) [inline src/merkle/store.rs:400]: This formatting makes it a bit difficult to follow. Maybe update to something like:
```Rust
assert_eq!(
    store.get_node(mtree.root(), NodeIndex::new(mtree.depth(), 0)),
    Ok(LEAVES4[0]),
    "node 0 must be in the tree"
);
```
If this doesn't work, maybe introduce `depth` variable (set to `mtree_depth()`) to make `rustfmt` accept it?

File: src/merkle/store.rs (hunk 7 above)
Comment (hackaugusto) [reply src/merkle/store.rs:122]: done

File: src/merkle/store.rs [hunk 10]
```diff
@@ -110,20 +110,39 @@ impl MerkleStore {
...
+        root: Word,
+        index: NodeIndex,
+    ) -> Result<(Word, MerklePath), MerkleError> {
+        let mut hash: RpoDigest = root.into();
+        let mut path = Vec::new();
+
+        // Check the root is in the storage when called with `NodeIndex::root()`
+        self.nodes
+            .get(&hash)
+            .ok_or(MerkleError::RootNotInStorage(hash.into()))?;
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply src/merkle/store.rs:127]: what I was trying to say is that if the user gave `NodeIndex::root()` the loop below will not be executed, so this check is necessary only for that case, to make sure the `root` is in the storage.

[the remark this answers is not in the document]
Comment (hackaugusto) [reply src/merkle/store.rs:127]: updated the docs and changed the comment to:

```rust
// corner case: check the root is in the storage when called with index set to
// `NodeIndex::root()`
```

**Comment by hackaugusto** [comment]
> Also, since there are quite a few tests now, let's split them out into a separate file. So, we'd have src/merkle/store/mod.rs and src/merkle/store/tests.rs.

done

File: src/merkle/store/mod.rs [hunk 11]
```diff
@@ -108,22 +113,45 @@ impl MerkleStore {
         Ok(hash.into())
     }
 
-    /// Returns the path for the node at `index` rooted on the tree `root`.
+    /// Returns the leaf value at `index` and its opening to `root`.
```
Comment (bobbinth) [inline src/merkle/store/mod.rs:116]: Should this be "leaf value"? Technically we would be returning any node value at the specified index. So, maybe a better phrasing would be "Returns the node at the specified `index` and its opening to the `root`"?

File: src/merkle/store.rs (hunk 4 above)
Comment (bobbinth) [reply src/merkle/store.rs]: Ah - maybe related to this https://github.com/0xPolygonMiden/crypto/issues/57?

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you! I left a couple of non-blocking comments inline.

File: src/merkle/mod.rs [hunk 12]
```diff
@@ -71,6 +72,7 @@ impl fmt::Display for MerkleError {
             InvalidEntriesCount(max, provided) => write!(f, "the provided number of entries is {provided}, but the maximum for the given depth is {max}"),
             NodeNotInSet(index) => write!(f, "the node indexed by {index} is not in the set"),
             NodeNotInStorage(hash, index) => write!(f, "the node {:?} indexed by {} and depth {} is not in the storage", hash, index.value(), index.depth(),),
+            RootNotInStorage(root) => write!(f, "the root {:?} is not in the storage", root),
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply src/merkle/mod.rs:75]: renamed

File: src/merkle/store/mod.rs (hunk 11 above)
Comment (hackaugusto) [reply src/merkle/store/mod.rs:116]: changed from `leaf` to `node`

File: src/merkle/store.rs (hunk 4 above)
Comment (hackaugusto) [reply src/merkle/store.rs]: seem related
